lib/floki/selector/pseudo_class.ex

```diff
@@ -152,6 +152,20 @@ defmodule Floki.Selector.PseudoClass do
     html_node.node_id in tree.root_nodes_ids
   end
 
+  def match_has?(tree, html_node, %__MODULE__{value: value} = pseudo_class) do
+    value
+    |> Enum.all?(fn inner_selector ->
+      html_node.children_nodes_ids
+      |> Enum.any?(fn id ->
+        child = Map.get(tree.nodes, id)
+
+        child |> is_struct(HTMLNode) and
+          (child |> Floki.Selector.match?(inner_selector, tree) or
+             match_has?(tree, child, pseudo_class))
+      end)
+    end)
```

Comment:
Just some style preferences:

```suggestion
    Enum.all?(value, fn inner_selector ->
     Enum.any?(html_node.children_nodes_ids, fn id ->
        child = Map.fetch!(tree.nodes, id)

        is_struct(child, HTMLNode) and
          (Floki.Selector.match?(child, inner_selector, tree) or
             match_has?(tree, child, pseudo_class))
      end)
    end)
```

1. the pipe operator is nice, but I prefer to avoid it when there is only one step to pipe.
2. we expect that the node is always present in the tree, so using `Map.fetch!/2` is more assertive.

Reply:
I'm curious why we need to filter only for `HTMLNode` here. Is there any problem with matching against all nodes?

Reply:
> I'm curious why we need to filter only for HTMLNode here.

Because the selectors inside `:has()` can only match on HTML elements (`HTMLNode`), not on `Text` or `Comment` nodes.